clients/src/main/java/org/apache/kafka/clients/KafkaClient.java

```diff
@@ -112,24 +111,21 @@
     public int inFlightRequestCount(String nodeId);
 
     /**
-     * Generate a request header for the next request
-     * 
-     * @param key The API key of the request
+     * Wake up the client if it is currently blocked waiting for I/O
      */
-    public RequestHeader nextRequestHeader(ApiKeys key);
+    public void wakeup();
 
     /**
-     * Generate a request header for the given API key
+     * Create a new ClientRequest.
      *
-     * @param key The api key
-     * @param version The api version
-     * @return A request header with the appropriate client id and correlation id
+     * @param nodeId the node to send to
+     * @param requestBuilder the request builder to use
+     * @param createdTimeMs the time in milliseconds to use as the creation time of the request
+     * @param expectResponse true iff we expect a response
+     * @param callback the callback to invoke when we get a response
+     * @return
      */
-    public RequestHeader nextRequestHeader(ApiKeys key, short version);
-
-    /**
-     * Wake up the client if it is currently blocked waiting for I/O
-     */
-    public void wakeup();
-
+    public ClientRequest newClientRequest(String nodeId, AbstractRequest.Builder requestBuilder,
```

Comment:
Hmm.  I think if it were just newRequest, I would wonder whether it was a new AbstractRequest or a new ClientRequest.  What do you think?